### PR TITLE
[Snappi] PFC code refactor for Snappi

### PR DESCRIPTION
**Pull request by developfast**
### Description of PR

Summary: Codebase for snappi based PFC tests were becoming very reliant on new parameters being introduced to write new tests, as well as significant chunks of repetitive code with no regard for which verification standards applied for which tests. The code refactor solves all these issues, and the code has been written to become very modular - allowing us to introduce new vars to tests without cluttering any single function. This is one part of the plan to refactor the entire snappi directory. 

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [X] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [] 201911
- [X] 202012
- [X] 202205

### CODE REVIEW

File: tests/common/snappi/traffic_generation.py [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+        """ Set flow port config values """
+        flow_port_config = base_flow_config["flow_port_config"]
+        flow_port_config[0][str(base_flow_config["tx_port_config"].peer_port)].append(int(prio))
+        flow_port_config[1][str(base_flow_config["rx_port_config"].peer_port)].append(int(prio))
+        base_flow_config["flow_port_config"] = flow_port_config
+
+    snappi_extra_params.base_flow_config = base_flow_config
+
+
+def generate_background_flows(testbed_config,
```
Comment (baiwei0427) [inline tests/common/snappi/traffic_generation.py:137]: Can we use a single function to generate both test flows and background flows?

File: tests/common/snappi/traffic_generation.py [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+def verify_background_flow(flow_metrics,
+                           bg_flow_name,
+                           bg_flow_rate_percent,
+                           data_flow_dur_sec,
+                           data_pkt_size,
+                           speed_gbps,
+                           tolerance,
+                           snappi_extra_params):
+    """
+    Verify background flow statistics
```
Comment (baiwei0427) [inline tests/common/snappi/traffic_generation.py:364]: What are expected behaviors for background flows?

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/common/snappi/traffic_generation.py (hunk 1 above)
Comment (developfast) [reply tests/common/snappi/traffic_generation.py:137]: I am keeping them separate in case in the future we want to only run test flows or background flows. I would also want one traffic gen function to generate exactly one type of flow.

File: tests/common/snappi/snappi_test_params.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            packet_capture_type (ENUM): packet capture type
+            packet_capture_file (str): packet capture file
+            packet_capture_ports (list): packet capture ports on ixia chassis
+            base_flow_config (dict): base flow configuration
+            test_tx_frames (list): number of test frames transmitted for priorities to test ex. [2000, 3000]
+                for priorities 3 and 4
+        """
+        self.headroom_test_params = None
+        self.pfc_pause_src_mac = None
+        self.set_pfc_class_enable_vec = True
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/snappi_test_params.py:27]: corrected

File: tests/common/snappi/snappi_test_params.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+from tests.common.snappi.common_helpers import packet_capture
+
+
+class SnappiTestParams():
+    def __init__(self):
+        """
+        Initialize the SnappiTestParams class
+
+        Params:
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/snappi_test_params.py:13]: Done

File: tests/common/snappi/common_helpers.py [hunk 5]
```diff
@@ -722,3 +723,45 @@ def get_egress_queue_count(duthost, port, priority):
...
+    """
+
+    cap = testbed_config.captures.capture(name=capture_name if capture_name else "PacketCapture")[-1]
+    cap.port_names = [testbed_config.ports[port_id].name]
+    cap.format = cap.PCAP
+
+    if capture_type == packet_capture.IP_CAPTURE:
+        # Capture IP packets
+        ip_filter = cap.filters.custom()[-1]
+        ip_filter.value = '40'
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/common_helpers.py:762]: Done

**Comment by developfast** [comment]
Have addressed all comments - pls approve @baiwei0427

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/common/snappi/common_helpers.py [hunk 6]
```diff
@@ -722,3 +723,46 @@ def get_egress_queue_count(duthost, port, priority):
...
+    cap = testbed_config.captures.capture(name=capture_name if capture_name else "PacketCapture")[-1]
+    cap.port_names = [testbed_config.ports[port_id].name]
+    cap.format = cap.PCAP
+
+    if capture_type == packet_capture.IP_CAPTURE:
+        # Capture IP packets
+        ip_filter = cap.filters.custom()[-1]
+        # Version for IPv4 packets is "4" which has to be in the upper 4 bits of the first byte, hence filter is 0x40
+        ip_filter.value = '40'
+        ip_filter.offset = 14
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/common_helpers.py:763]: No

File: tests/common/snappi/common_helpers.py [hunk 7]
```diff
@@ -722,3 +723,46 @@ def get_egress_queue_count(duthost, port, priority):
...
+        ip_filter.mask = '0f'  # Mask is 0x0f to only match the upper 4 bits of the first byte which is the version
+    elif capture_type == packet_capture.PFC_CAPTURE:
+        # Capture PFC packets
+        pfc_filter = cap.filters.custom()[-1]
+        pfc_filter.value = '8808'  # PFC pause MAC control code is 0x8808, hence this is the filter value
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/common_helpers.py:768]: This filter is slightly incorrect - I will remove it

File: tests/common/snappi/traffic_generation.py [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+        test_flow.metrics.enable = True
+        test_flow.metrics.loss = True
+
+        """ Set flow port config values """
+        dut_port_config = base_flow_config["dut_port_config"]
+        dut_port_config[0][str(base_flow_config["tx_port_config"].peer_port)].append(int(prio))
+        dut_port_config[1][str(base_flow_config["rx_port_config"].peer_port)].append(int(prio))
+        base_flow_config["dut_port_config"] = dut_port_config
+
+    snappi_extra_params.base_flow_config = base_flow_config
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/traffic_generation.py:143]: Because background flows make no changes to the dut port config which only contains lossless prios

File: tests/common/snappi/traffic_generation.py [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+            continue
+
+        tx_frames = metric.frames_tx
+        rx_frames = metric.frames_rx
+
+        exp_bg_flow_rx_pkts = bg_flow_rate_percent / 100.0 * speed_gbps \
+            * 1e9 * bg_flow_dur_sec / 8.0 / bg_flow_pkt_size
+        deviation = (rx_frames - exp_bg_flow_rx_pkts) / float(exp_bg_flow_rx_pkts)
+
+        if snappi_extra_params.headroom_test_params is None:
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/traffic_generation.py:399]: Good catch

File: tests/common/snappi/traffic_generation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (baiwei0427) [reply tests/common/snappi/traffic_generation.py:143]: > Because background flows make no changes to the dut port config which only contains lossless prios

We may also run background flows on lossless priorities, e.g., run test flows on priority 4, and background flows on priority 3.

File: tests/common/snappi/traffic_generation.py [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+    Args:
+        testbed_config (obj): testbed L1/L2/L3 configuration
+        test_flow_name (str): name of test flow
+        test_flow_prio_list (list): list of test flow priorities
+        test_flow_rate_percent (int): rate percentage of test flows
+        test_flow_dur_sec (int): duration of test flows
+        test_flow_delay_sec (int): delay of test flows in seconds
+        test_flow_pkt_size (int): packet size of test flows
+        prio_dscp_map (dict): priority to DSCP mapping
+        snappi_extra_params (obj): snappi extra parameters
```
Comment (baiwei0427) [inline tests/common/snappi/traffic_generation.py:107]: Can you specify the type of this variable should be SnappiTestParams?

File: tests/common/snappi/traffic_generation.py [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+
+    Args:
+        flow_metrics (list): per-flow statistics
+        test_flow_name (str): name of the test flow
+        test_flow_rate_percent (int): test flow rate in percentage
+        test_flow_dur_sec (int): test flow duration in second
+        test_flow_pkt_size (int): test packet size in bytes
+        speed_gbps (int): speed of the port in Gbps
+        tolerance (float): tolerance for test flow deviation
+        test_flow_pause (bool): whether test flow is paused
```
Comment (baiwei0427) [inline tests/common/snappi/traffic_generation.py:427]: is expected to be paused.

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/common/snappi/common_helpers.py [hunk 12]
```diff
@@ -722,3 +723,42 @@ def get_egress_queue_count(duthost, port, priority):
     total_pkts = raw_out.split()[2]
     total_bytes = raw_out.split()[3]
     return int(total_pkts.replace(',', '')), int(total_bytes.replace(',', ''))
+
+
+class packet_capture(Enum):
+    """
+    ENUM of packet capture settings
+    NO_CAPTURE - No capture
+    PFC_CAPTURE_EGRESS - PFC capture enabled
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/common_helpers.py:732]: that was incorrect naming - changed to a generic PFC_CAPTURE

File: tests/common/snappi/traffic_generation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi/traffic_generation.py:143]: dut port config dict is solely used in the verify test flows func so the background flows are irrelevant

**Review by baiwei0427 — APPROVED** [review]

**Review by neethajohn — APPROVED** [review]

**Comment by wangxin** [comment]
Branch 202305 was branched out before this PR was merged. Removing 202305 related labels.
